proof/refine/AARCH64/ArchKHeap_R.thy

```diff
@@ -0,0 +1,413 @@
+  apply (rule hoare_lift_Pf [where f="ksArchState"])
+   apply (rule hoare_lift_Pf [where f="ksIdleThread"])
+    apply (rule hoare_lift_Pf [where f="irq_node'"])
+     apply (rule hoare_lift_Pf [where f="gsMaxObjectSize"])
+      apply (wp ctes hoare_vcg_const_Ball_lift arch idle irqn maxObj)+
+  done
+
+lemma valid_arch_state_lift':
+  assumes typs: "\<And>T p P. f \<lbrace>\<lambda>s. P (typ_at' T p s)\<rbrace>"
+  assumes arch: "\<And>P. f \<lbrace>\<lambda>s. P (ksArchState s)\<rbrace>"
+  assumes vcpu: "\<And>P p. f \<lbrace>\<lambda>s. P (ko_wp_at' (is_vcpu' and hyp_live') p s)\<rbrace>"
+  shows "f \<lbrace>valid_arch_state'\<rbrace>"
+  apply (simp add: valid_arch_state'_def valid_asid_table'_def vspace_table_at'_defs)
+  apply (wp_pre, wps arch)
+   apply (wpsimp wp: hoare_vcg_all_lift hoare_vcg_const_imp_lift vcpu[unfolded pred_conj_def]
+                 split: option.split)
+  apply (clarsimp simp: pred_conj_def)
+  done
+
+lemma set_ntfn_hyp[wp]:
+  "setNotification ptr val \<lbrace>\<lambda>s. P (ko_wp_at' (is_vcpu' and hyp_live') p s)\<rbrace>"
+  unfolding setNotification_def
+  by (wpsimp wp: setObject_ko_wp_at simp: gen_objBits_simps, rule refl, simp)
+     (clarsimp simp: is_vcpu'_def ko_wp_at'_def obj_at'_def)
+
+lemma set_ep_hyp[wp]:
+  "setEndpoint ptr val \<lbrace>\<lambda>s. P (ko_wp_at' (is_vcpu' and hyp_live') p s)\<rbrace>"
+  unfolding setEndpoint_def
+  by (wpsimp wp: setObject_ko_wp_at simp: objBits_simps', rule refl, simp)
+     (clarsimp simp: is_vcpu'_def ko_wp_at'_def obj_at'_def)
+
+lemma idle_is_global[KHeap_R_assms, intro!]:
+  "ksIdleThread s \<in> global_refs' s"
+  by (simp add: global_refs'_def)
+
+end
+
+interpretation KHeap_R?: KHeap_R
+proof goal_cases
+  interpret Arch .
+  case 1 show ?case by (intro_locales; (unfold_locales; fact KHeap_R_assms)?)
+qed
+
+context Arch begin arch_global_naming
+
+named_theorems KHeap_R_assms_2
+
+lemmas setEndpoint_valid_globals[KHeap_R_assms_2, wp]
+  = valid_global_refs_lift'[OF set_ep_ctes_of set_ep_arch'
+                               setEndpoint_it setEndpoint_ksInterruptState]
+
+lemma set_ntfn_global_refs'[KHeap_R_assms_2, wp]:
+  "\<lbrace>valid_global_refs'\<rbrace> setNotification ptr val \<lbrace>\<lambda>_. valid_global_refs'\<rbrace>"
+  by (rule valid_global_refs_lift'; wp)
+
+crunch setEndpoint, setNotification
+  for valid_arch'[wp]: valid_arch_state'
+  (wp: valid_arch_state_lift')
+
+lemmas [KHeap_R_assms_2] = setEndpoint_valid_arch' setNotification_valid_arch'
+
+lemmas setObject_typ_ats[wp] = typ_at_lifts[OF setObject_typ_at']
+
+lemmas doMachineOp_typ_ats[wp] = typ_at_lifts[OF doMachineOp_typ_at']
+
+lemmas setEndpoint_typ_ats[wp] = typ_at_lifts[OF setEndpoint_typ_at']
+
+end
+
+interpretation KHeap_R_2?: KHeap_R_2
+proof goal_cases
+  interpret Arch .
+  case 1 show ?case by (intro_locales; (unfold_locales; fact KHeap_R_assms_2)?)
+qed
+
+arch_requalify_facts setObject_other_corres (* FIXME arch-split: move *)
```

Comment:
Should this move be done now or is it being left until we've arch-split up to a file where it makes sense to be moved to?

Reply:
The latter, but looking at it now I'm not sure such a place will appear, as the next one that pulls it in is SubMonad_R (and also Init_R). Hmm.

Reply:
Changed the comment on this, and added the setObject pspace_in_kernel_mappings' lemma that has the same curse.

Reply:
Works for me.